docs/INTRO.md

```diff
@@ -0,0 +1,462 @@
+
+  >>> f # x # y
+  f x y
+-}
+(#) :: Term s (a :--> b) -> Term s a -> Term s b
+infixl 8 #
+
+{- |
+  Low precedence infixr function application, to be used like
+  `$`, in combination with `#`. e.g.:
+
+  >>> f # x #$ g # y # z
+  f x (g y z)
+-}
+(#$) :: Term s (a :--> b) -> Term s a -> Term s b
+infixr 0 #$
+```
+
+The types of each operator match our intuition. Applying a lambda `Term` to a `Term` (tagged with the `PType` of the domain of the lambda) produces a `Term` (tagged with the `PType` of the codomain.).
+
+### Strictness and Laziness; Delayed Terms and Forcing
+
+Plutarch, like UPLC, is strict by default; this is in contrast to Haskell, which is nonstrict. In practice, this means that calling a function in Plutarch evaluates _all_ arguments to the Plutarch lambda `Term` beforehand.
+
+> Note: the below example does not correspond precisely to the implementation of `pif` or `pif'`; it is for didactic purposes only
+
+This behavior may be undesirable, for example, when one of two `Term`s are branched upon within an `if` statement. In Plutarch, `pif'` is strict and `pif` is lazy. A strict `pif'` could look something like 
+
+```hs
+{- | Haskell-level `if` function between `Term`s. This  function is nonstrict;
+only a single branch is evaluated (since this is a Haskell function).
+Unfortunately, it _doesn't_ translate to a non-strict Plutarch lambda, 
+as we will see below.
+-}
+hif' :: Term s PBool -> Term s a -> Term s b -> Term s c
+hif' cond whenTrue whenFalse = pmatch cond $ \case
+  PTrue -> whenTrue
+  PFalse -> whenFalse
+
+{- | Strict Plutarch-level `if`. The arguments _are_ evaluated
+(by default), meaning that both branches are evaluated.
+pif' :: forall {s :: S} {b :: PType}. Term s (PBool :--> (b :--> (b :--> b)))
+pif' = plam hif
+```
+
+A strict `if` is undesirable for the obvious reason: we only follow one branch at runtime, so it doesn't make sense to evaluate both before examining the `PBool` value to which we apply `pif`.
+
+To avoid this, we use `pdelay` to create a "delayed `Term`." `pdelay` wraps the `PType` tag of a term, overriding the default strict behavior and indicating that the term should _not_ be evaluated immediately. `pdelay` has the following type:
+
+```hs
+pdelay :: Term s a -> Term s (PDelayed a)
+```
+
+A delayed term evaluates when it is _forced_ using the `pforce` function. Forcing a term strips the `PDelayed` wrapper:
+
+```hs
+pforce :: Term s (PDelayed a) -> Term s a 
+```
+
+Thus, if we wanted a lazy `pif", we could do the following:
+
+```hs
+{- | Haskell-level `if` function between `Term`s. This time
+it _does_ translate to a lazy Plutarch lambda.
+-}
+hif :: Term s PBool -> Term s a -> Term s a -> Term s a
+hif cond whenTrue whenFalse = pforce $ pif' # cond # pdelay whenTrue # pdelay whenFalse
+
+{- | Plutarch-level `if`. We delay the arguments in `hif` so they are not applied
+until forced.
+-}
+pif :: forall {s :: S} {b :: PType}. Term s (PBool :--> (b :--> (b :--> b)))
+pif = plam hif
```

Comment:
@peter-mlabs  One thing that I didn't notice but might still be helpful for your understanding. This `pif` has the same issues as the previous one we talked about. Its type is `Term s (PBool :--> b :--> b :--> b)`. All of these types are immediately evaluated before the body of `hif` is even entered - so we're back to square one. No use delaying inside the body.

You can't make lazy Plutarch level functions ergonomically, at all. And there's no point in doing so. A lazy Plutarch level function would have a signature like: `Term s (PBool :--> PDelay b :--> PDelay b :--> b)`

Reply:
Yeah, after spending some time working with it, this is all starting to sink in. Thanks for the help and feedback!